src/main/java/updater/JarUpdater.java

```diff
@@ -0,0 +1,302 @@
+     */
+    private boolean prepareTargetJarFile(File targetJarFile) {
+        if (targetJarFile.getParentFile() != null &&
+                !createDirectories(targetJarFile.getParentFile().getAbsoluteFile())) {
+            log("Failed to make directories of target file.");
+            return false;
+        }
+
+        if (targetJarFile.exists() && !makeJarBackup(targetJarFile)) {
+            log("Failed to make backup.");
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Makes back up of a JAR file.
+     *
+     * In some platforms (Windows in particular), JAR file cannot be modified if it was executeded and
+     * the process has not ended yet. As such, we will make several tries with wait in making backup.
+     *
+     * @param jarFile Jar file to be backed up
+     * @return true if backup is created, false otherwise
+     */
+    private boolean makeJarBackup(File jarFile) {
+        log("Making JAR backup");
+
+        for (int i = 0; i < MAX_RETRY; i++) {
+            if (moveFileToBackup(jarFile)) {
+                return true;
+            }
+
+            log("Failed to make backup. Might be due to original JAR still in use.");
+            try {
+                log("Wait for a while before trying again.");
+                Thread.sleep(WAIT_TIME);
+            } catch (InterruptedException e) {
+                log("Failed to wait for a while");
+            }
+        }
+
+        log("Jar file cannot be backed up. Most likely is in use by another process.");
+        return false;
+    }
+
+    /**
+     * Moves file to its backup file (the original file will be removed)
+     * @param file file to be moved to its backup
+     * @return true if move is successful, false otherwise
+     */
+    private boolean moveFileToBackup(File file) {
+        log("Moving file as backup");
+        String backupFilename = getBackupFilename(file.getName());
+        File backup = new File(file.getParent(), backupFilename);
+        return moveFile(file.toPath(), backup.toPath());
+    }
+
+    /**
+     * Moves a backup file to its original file
+     * @param file the original file to be recovered
+     * @return true if backup is restored, false otherwise
+     */
+    private boolean restoreBackup(File file) {
+        log("Restoring backup");
+        String backupFilename = getBackupFilename(file.getName());
+        File backup = new File(file.getParent(), backupFilename);
+        return moveFile(backup.toPath(), file.toPath());
+    }
+
+    /**
+     * Removes the backup of a file
+     * @param file the file which backup should be removed
+     */
+    private void removeBackup(File file) {
+        log("Removing backup");
+        String backupFilename = getBackupFilename(file.getName());
+        File backup = new File(file.getParent(), backupFilename);
+        if (backup.exists() && !backup.delete()) {
+            log("Failed to remove backup");
+        }
```

Comment:
I know this code works, but it is written in a misleading way. Make the logic more straight-forward, even if it makes the code more verbose. Can also extract a more general `deleteFile` method.